### ISSUE TITLE
changing anchors and image min max sides during fine tuning / inference

### ISSUE THREAD

**Issue by tonmoyborah** [fizyr/keras-retinanet#1240]
Hi, this might not belong here but I'd still like to ask if it's possible to change anchor shapes and sizes when fine tuning my custom model on new dataset. Do we have to maintain the same anchor config throughout all training and fine tuning? Also the same thing about image-min-side and image-max-side. Does this make sense? 
Anyway it throws `IndexError` if I change it without any other change.

**Comment by tonmoyborah** [comment]
I am noticing that if I have a trained model with default settings, now I use this model for weights and try to train it on new data but with added flags `--image-min-side 700` and `--image-max-side 700`, essentially fine tuning my own model, training runs fine but results get weird. All boxes have <0.2 score and are clustered to top right area of the image.

**Comment by hgaiser** [comment]
Anchor shapes and sizes correlate with how the feature pyramid network is setup, why do you want to change them? Maybe you intended to change scales and ratios?

**Comment by tonmoyborah** [comment]
Yes, sorry poor description, I was talking about scales and ratios. But my main concern is how changing image size during transfer learning works. But I guess due to how feature map convolution to weight matrix works, this essentially starts everything anew as activations are messed up. 
This seems to be not working only when I try fine tuning my own custom model over again with smaller image sizes. If I try to use the coco_best pretrained model directly with smaller image sizes, it starts off fine as usual.

**Comment by hgaiser** [comment]
> Yes, sorry poor description, I was talking about scales and ratios. But my main concern is how changing image size during transfer learning works. But I guess due to how feature map convolution to weight matrix works, this essentially starts everything anew as activations are messed up.

Yes, I think you're correct

> This seems to be not working only when I try fine tuning my own custom model over again with smaller image sizes. If I try to use the coco_best pretrained model directly with smaller image sizes, it starts off fine as usual.

How long have you been training? Did you freeze any layers? Is the loss decreasing?

**Comment by tonmoyborah** [comment]
Yes, the training is working fine but the issue is, for the same training set, and image sizes 700, 700:
1. If I use coco pre-trained model for weights, the loss starts off at something like 1.9 and 0.9 respectively (regress and classification). When I test this, even after 3 epochs, the results are decent.
2. If I use my old trained model (which actually works great for images similar to this new training set), the loss starts at a much higher value. After about 5 epochs (tested even for 15 epochs, 25000 steps) the boxes are all weirdly occurring towards top left of the image and with very low confidence (0.01 to 0.3)
I tried freezing and unfreezing the backbone in both above cases without much different results. I don't know if there is a user friendly option to freeze selected layers in this repository.

**Comment by hgaiser** [comment]
One possibility, though I wouldn't expect it to happen, is that it is pushed into a state where it is difficult to get out of. In those cases it requires a lot of training to get out of the state. If finetuning from COCO works, I would simply go with that route :)

If your issue is resolved, please close this issue.

**Comment by tonmoyborah** [comment]
Thanks, I'll investigate further

**Comment by tonmoyborah** [comment]
After further investigation, I realized that I cannot actually make the `--image-min-side` and `--image-max-side` arguments work for me even with pretrained model. Has anyone tested these arguments during training with the default coco pretrained model provided (resnet50_coco_best_v2.1.0.h5).
My arguments while training are like so:

    --epochs 5 --batch-size 2 --random-transform --backbone resnet50 --workers 4 --steps 60000 --lr 1e-04 --image-min-side 700 --image-max-side 700 --weights /path/to/resnet50_coco_best_v2.1.0.h5 --snapshot-path /path/to/model.h5 --tensorboard-dir /path --config /path/to/config.ini csv path/to/annotations_csv.txt /path/to/class_names.txt

After this training, when I use evaluate script, mAP is around 0.004 with arguments

    --backbone resnet50 --gpu 0 --image-min-side 700 --image-max-side 700 --config /path/to/config.ini --convert-model csv path/to/annotations_csv.txt /path/to/class_names.txt /path/to/trained_model/resnet50_csv_05.h5

When I use the script to visualize the detections, all boxes are clustered to top left corner with scores around 0.1

**Comment by hgaiser** [comment]
>  Do we need to train from scratch and cannot use the given coco pretrained model with changed image sizes?

We do that all the time, should work fine! I don't see anything wrong with your flags.

**Comment by tonmoyborah** [comment]
Ok, thanks for the confirmation. Closing this since probably there is a problem at my end
